### PR TITLE
Generation evaluation

### PR DESCRIPTION
**Pull request by dgbaenar**
* Generation in batch as endpoint
* Evaluate generation as endpoint

### CODE REVIEW

File: app/evaluation/generation/correctness.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+        except KeyError as e:
+            logging.error(f"Missing key in JSON response: {e}")
+            return {"score": None, "passing": None, "reasoning": "Incomplete JSON response"}
+
+        except Exception as e:
+            logging.error(f"An error occurred: {e}")
+            return {"score": None, "passing": None, "reasoning": "An unexpected error occurred"}
+
+    def run_batch_evaluation(self,
```
Comment (mats-claassen) [inline app/evaluation/generation/correctness.py]: Is this batch evaluation? It looks like it calls OpenAI for each single element

Comment (mats-claassen) [reply app/evaluation/generation/correctness.py]: Maybe we should change the name of the function for clarity. For example, "evaluate_dataset"

File: app/evaluation/generation/faithfulness.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            except Exception as e:
+                print(f"Error encountered: {e}. Saving progress and exiting.")
+                raise
+
+        # Load the results back into a DataFrame and concatenate with the original
+        results_df = pd.read_csv(output_file)
+
+        total_questions = len(results_df)
+        faithfulness_relevancy = round(results_df["relevancy"].sum(
+        ) / total_questions, 2)
```
Comment (mats-claassen) [inline app/evaluation/generation/faithfulness.py]: Some lines like this ones are split into multiple lines even though they are very short. The line-length is set to 130 in .ruff.toml. This complete line has only 91 chars. Do you have a different local setting that overrides the ruff configuration?

File: app/routes/evaluation_endpoints.py [hunk 3]
```diff
@@ -88,6 +101,173 @@ async def evaluate_retrieval(
...
+    max_tokens: int = Form(400),
+    limit: int = Form(None),
+    # Can be 'correctness' or 'faithfulness'
+    evaluation_type: str = Form(...),
+    query_column: str = Form("openai_query"),
+    reference_answer_column: str = Form(
+        "openai_answer"),
+    generated_answer_column: str = Form("response"),
+    resource_id_column: str = Form(
+        "resource_id_source"),
```
Comment (mats-claassen) [inline app/routes/evaluation_endpoints.py]: Here for example, this line would be only 58 chars

File: app/routes/evaluation_endpoints.py [hunk 4]
```diff
@@ -88,6 +101,173 @@ async def evaluate_retrieval(
...
+
+        if evaluation_type == "correctness":
+            evaluator = CorrectnessEvaluator(
+                openai_api_key, openai_model, threshold=correctness_threshold, max_tokens=max_tokens)
+            evaluation_metrics = evaluator.run_batch_evaluation(df=df, output_file=output_file,
+                                                                query_column=query_column,
+                                                                reference_answer_column=reference_answer_column,
+                                                                generated_answer_column=generated_answer_column,
+                                                                resource_id_column=resource_id_column)
+        elif evaluation_type == "faithfulness":
```
Comment (mats-claassen) [inline app/routes/evaluation_endpoints.py]: Correctness and faithfulness should be tracked in the same experiment

Comment (dgbaenar) [reply app/routes/evaluation_endpoints.py]: Fixed

File: app/routes/evaluation_endpoints.py (hunk 3 above)
Comment (dgbaenar) [reply app/routes/evaluation_endpoints.py]: Fixed

File: app/evaluation/generation/faithfulness.py (hunk 2 above)
Comment (dgbaenar) [reply app/evaluation/generation/faithfulness.py]: Fixed, I was doing it with autopep8 using ctrl+shift+i in VS code. I was running the checker and not the ruff formater.

File: app/evaluation/generation/correctness.py (hunk 1 above)
Comment (dgbaenar) [reply app/evaluation/generation/correctness.py]: Fixed.

**Pull request merged**